include/utils.hpp

```diff
@@ -0,0 +1,137 @@
+#pragma once
+#include <fmt/format.h>
+
+#include "mango_v3.hpp"
+
+namespace mango_v3 {
+// quoteFree, quoteLocked, baseFree, baseLocked
+auto splitOpenOrders(const serum_v3::OpenOrders& openOrders) {
+  const auto quoteFree =
+      openOrders.quoteTokenFree + openOrders.referrerRebatesAccrued;
+  const auto quoteLocked =
+      (openOrders.quoteTokenTotal - openOrders.quoteTokenFree);
+  const auto baseFree = openOrders.baseTokenFree;
+  const auto baseLocked = openOrders.baseTokenTotal - openOrders.baseTokenFree;
+  return std::make_tuple(quoteFree, quoteLocked, baseFree, baseLocked);
+}
+auto getUnsettledFunding(const PerpAccountInfo& accountInfo,
+                         const PerpMarketCache& perpMarketCache) {
+  if (accountInfo.basePosition < 0) {
+    return accountInfo.basePosition *
+           (perpMarketCache.short_funding - accountInfo.shortSettledFunding);
+  } else {
+    return accountInfo.basePosition *
+           (perpMarketCache.long_funding - accountInfo.longSettledFunding);
+  }
+}
+// Return the quote position after adjusting for unsettled funding
+auto getQuotePosition(const PerpAccountInfo& accountInfo,
+                      const PerpMarketCache& perpMarketCache) {
+  return accountInfo.quotePosition -
+         getUnsettledFunding(accountInfo, perpMarketCache);
+}
+/**
+ * Return weights corresponding to health type;
+ * Weights are all 1 if no healthType provided
+ * @return
+ * <spotAssetWeight, spotLiabWeight,perpAssetWeight, perpLiabWeight>
+ */
+auto getMangoGroupWeights(const MangoGroup& mangoGroup, uint64_t marketIndex,
+                          HealthType healthType = HealthType::Unknown) {
+  if (healthType == HealthType::Maint) {
+    return std::make_tuple(mangoGroup.spotMarkets[marketIndex].maintAssetWeight,
+                           mangoGroup.spotMarkets[marketIndex].maintLiabWeight,
+                           mangoGroup.perpMarkets[marketIndex].maintAssetWeight,
+                           mangoGroup.perpMarkets[marketIndex].maintLiabWeight);
+  } else if (healthType == HealthType::Init) {
+    return std::make_tuple(mangoGroup.spotMarkets[marketIndex].initAssetWeight,
+                           mangoGroup.spotMarkets[marketIndex].initLiabWeight,
+                           mangoGroup.perpMarkets[marketIndex].initAssetWeight,
+                           mangoGroup.perpMarkets[marketIndex].initLiabWeight);
+  } else {
+    return std::make_tuple(i80f48(1.0), i80f48(1.0), i80f48(1.0), i80f48(1.0));
+  }
+}
+auto nativeI80F48ToUi(i80f48 amount, uint8_t decimals) {
+  return amount / pow(10, decimals);
```

Comment:
which pow is this and does it have enough precision? 12 decimals is the largest i have seen

Reply:
i think using 10LL instead would be more correct

Reply:
just looked up pow, it's apparently using float/double under the hood, so 10.0L would be correct